src/hotspot/share/opto/arraycopynode.cpp

```diff
@@ -294,9 +294,20 @@ bool ArrayCopyNode::prepare_array_copy(PhaseGVN *phase, bool can_reshape,
     uint header = arrayOopDesc::base_offset_in_bytes(dest_elem);
 
     src_offset = Compile::conv_I2X_index(phase, src_offset, ary_src->size());
+    if (src_offset->is_top()) {
+      // Offset is out of bounds (the ArrayCopyNode will be removed)
+      return false;
+    }
     dest_offset = Compile::conv_I2X_index(phase, dest_offset, ary_dest->size());
-    if (src_offset->is_top() || dest_offset->is_top()) {
+    if (dest_offset->is_top()) {
       // Offset is out of bounds (the ArrayCopyNode will be removed)
+      PhaseIterGVN* igvn = phase->is_IterGVN();
+      if (can_reshape) {
+        if (!src_offset->is_top()) {
```

Comment:
The ifs can be merged.

Reply:
And looking at this again, `src_offset` can't be top, right? We just checked it above.